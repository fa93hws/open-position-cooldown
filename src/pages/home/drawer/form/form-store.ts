import type { IntentionSchema } from '@services/intention/intention';
import type { InputStore } from '@ui/text-input/input-store';
import { UnreachableException } from '@utils/unreachable-exception';
import type { BasicInfoStore } from './basic-info/basic-info-store';
import type { ReasonStore } from './reason/reason-store';
import type { StrategyStore } from './strategy/strategy-store';

const enum ValidateResult {
  BASIC_INFO_ERROR,
  PRICE_EXPLAIN_ERROR,
  REASON_ERROR,
  STRATEGY_ERROR,
  NO_ERROR,
}

export class FormStore {
  constructor(
    readonly basicInfoStore: BasicInfoStore,
    readonly reasonStore: ReasonStore,
    readonly priceExplainStore: InputStore,
    readonly strategyStore: StrategyStore,
    private readonly submitIntention: (
      intention: IntentionSchema,
    ) => Promise<void>,
  ) {}

  private validate(): ValidateResult {
    this.basicInfoStore.startValidate();
    this.reasonStore.startValidate();
    this.priceExplainStore.startValidate();
    this.strategyStore.startValidate();
    if (this.basicInfoStore.hasError) {
      return ValidateResult.BASIC_INFO_ERROR;
    }
    if (this.reasonStore.reasonErrors.includes(true)) {
      return ValidateResult.REASON_ERROR;
    }
    if (this.priceExplainStore.hasError) {
      return ValidateResult.PRICE_EXPLAIN_ERROR;
    }
    if (this.strategyStore.strategyErrors.includes(true)) {
      return ValidateResult.STRATEGY_ERROR;
    }
    return ValidateResult.NO_ERROR;
  }

  private toIntention(): IntentionSchema {
    return {
      basicInfo: {
        name: this.basicInfoStore.name,
        code: this.basicInfoStore.code,
        market: this.basicInfoStore.market,
      },
      reasons: this.reasonStore.reasonStrings,
      priceExplain: this.priceExplainStore.value,
      strategy: {
        initial: {
          quantity: this.strategyStore.initialQuantity,
          price: this.basicInfoStore.priceAsNum,
        },
        plan: this.strategyStore.strategyPlan,
        shitPrice: this.strategyStore.shitPrice,
      },
    };
  }

  async submit(alert: (msg: string) => void = window.alert) {
    const result = this.validate();
    switch (result) {
      case ValidateResult.NO_ERROR:
        await this.submitIntention(this.toIntention());
        return alert('submitted!'); // eslint-disable-line
      case ValidateResult.BASIC_INFO_ERROR:
        return alert('基本信息错误'); // eslint-disable-line
      case ValidateResult.REASON_ERROR:
        return alert('持有理由错误'); // eslint-disable-line
      case ValidateResult.PRICE_EXPLAIN_ERROR:
        return alert('价格解释错误'); // eslint-disable-line
      case ValidateResult.STRATEGY_ERROR:
        return alert('建仓策略错误'); // eslint-disable-line
      default:
        throw new UnreachableException(result);
    }
  }
}
