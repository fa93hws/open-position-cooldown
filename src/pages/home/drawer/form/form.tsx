import * as React from 'react';
import { Button, Box, Divider } from '@material-ui/core';
import { makeStyles, withTheme, WithTheme } from '@material-ui/core/styles';

import { designWidth } from '@styles/styles';
import { createBasicInfo } from './basic-info/basic-info';
import { createPriceExplain } from './price/price';
import { createReasonSection } from './reason/reason';

type FormProps = {
  onSubmit(): void;
  BasicInfo: React.ComponentType;
  PriceExplain: React.ComponentType;
  Reason: React.ComponentType;
};

const useStyles = makeStyles((theme) => ({
  container: {
    width: designWidth[0],
    margin: '0 auto',
    [theme.breakpoints.up('sm')]: {
      width: designWidth[1],
    },
  },
}));

export const Form = React.memo(
  withTheme((props: FormProps & WithTheme) => {
    const onSubmit = (e: React.FormEvent) => {
      e.preventDefault();
      props.onSubmit();
    };
    const styles = useStyles(props.theme);
    return (
      <Box
        component="form"
        onSubmit={onSubmit}
        display="flex"
        flexDirection="column"
        height="100%"
      >
        <Box padding={2} flex={1} className={styles.container}>
          <props.BasicInfo />
          <Divider />
          <Box mt={1}>
            <props.Reason />
          </Box>
          <Divider />
          <Box mt={1}>
            <props.PriceExplain />
          </Box>
        </Box>
        <Button fullWidth variant="contained" color="primary" type="submit">
          买！买！买！
        </Button>
      </Box>
    );
  }),
);

export function createForm() {
  const [BasicInfoComponent, basicInfoStore] = createBasicInfo();
  const [PriceExplainComponent, priceExplainStore] = createPriceExplain();
  const [ReasonComponent, reasonStore] = createReasonSection();

  const onSubmit = () => {
    basicInfoStore.startValidate();
    priceExplainStore.startValidate();
    reasonStore.startValidate();
  };
  return React.memo(() => (
    <Form
      onSubmit={onSubmit}
      BasicInfo={BasicInfoComponent}
      PriceExplain={PriceExplainComponent}
      Reason={ReasonComponent}
    />
  ));
}
