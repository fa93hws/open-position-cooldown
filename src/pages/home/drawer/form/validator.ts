export function nonEmpty(val: string) {
  return val.length > 0;
}

export function mustBeNumber(val: string) {
  return /^-?[0-9\\.]+$/.test(val);
}
