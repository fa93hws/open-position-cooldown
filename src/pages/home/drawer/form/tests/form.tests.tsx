import * as React from 'react';
import { shallow } from 'enzyme';

import { createForm, Form } from '../form';

describe('Form', () => {
  const BasicInfo = () => <div>BasicInfo</div>;
  const PriceExplain = () => <div>PriceExplain</div>;
  const Reason = () => <div>Reason</div>;
  const StrategyPanel = () => <div>StrategyPanel</div>;
  it('renders as follow', () => {
    expect(
      <Form
        onSubmit={jest.fn()}
        BasicInfo={BasicInfo}
        PriceExplain={PriceExplain}
        Reason={Reason}
        StrategyPanel={StrategyPanel}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('can be mounted', () => {
    const Component = createForm(jest.fn());
    expect(() => shallow(<Component />)).not.toThrow();
  });
});
