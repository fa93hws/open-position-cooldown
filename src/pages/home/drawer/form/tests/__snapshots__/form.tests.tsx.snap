// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders as follow 1`] = `
<form
  class="MuiBox-root MuiBox-root-3"
>
  <div
    class="MuiBox-root MuiBox-root-4 makeStyles-container-1"
  >
    <div>
      BasicInfo
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-5"
    >
      <div>
        Reason
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-6"
    >
      <div>
        PriceExplain
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-7"
    >
      <div>
        StrategyPanel
      </div>
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root makeStyles-button-2 MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      买！买！买！
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</form>
`;
